Warhead refactor part 3: Add WarheadArgs

reaperrr opened the pull request:
Adds `WarheadArgs` to simplify warhead code and make it easier to add more arguments in the future without having to update the warheads and their methods each time.

On the code side, the main concern is whether the initial set of args is sufficient and conceptually sound.
On the testing side, you only need to check whether this causes any regressions (which it shouldn't, but you never know).

reaperrr committed:
Introduce WarheadArgs

- Passes additional arguments to warheads on impact
- Uses that to reduce parameter count of DoImpact by 1

GraionDilach commented on OpenRA.Game/GameRules/WeaponInfo.cs:
```diff
@@ -33,6 +33,16 @@ public class ProjectileArgs
+	{
+		public WeaponInfo Weapon;
+		public int[] DamageModifiers;
+		public int LastProjectileFacing;
```
You're a hero.

GraionDilach commented:
Something I would propose to a followup: the last owner of the firer, to fix that issue when weapons fired by already-dead units explode without an owner.

reaperrr commented:
Updated. Streamlined this a bit and removed `LastProjectileFacing`, but don't worry @GraionDilach I have another follow-up that actually takes things one step further, using `WRot` for better forward-compatibility with the int-to-WAngle migration `Facing` will eventually get, as well as `Pitch` support (vertical impact angle).
This should make full directional armor support (front, rear, sides, top and bottom) much easier to implement.

reaperrr commented:
Rebased and decoupled from #17014.

@pchote This PR and #17280 are the ones I actually want most urgently (for my downstream directional armor implementation, mostly), so I'd prefer if you focused on this and #17280 over further discussion/review of #17014 for now.

pchote commented on OpenRA.Game/GameRules/WeaponInfo.cs:
```diff
@@ -177,17 +186,34 @@ public bool IsValidAgainst(FrozenActor victim, Actor firedBy)
 		}
+
+		/// <summary>Applies all the weapon's warheads to the target. Only use for projectile-less, special-case impacts.</summary>
+		public void Impact(Target impactTarget, Actor firedBy, IEnumerable<int> damageModifiers)
```
All but two uses of this overload call it as `Impact(Target.FromPos(somePos), someActor, Enumerable.Empty<int>())`. What do you think about changing this to `Impact(WPos pos, Actor firedBy)` and having the other two (`IonCannon`, `TeslaZap`)  construct their own `WarheadArgs` like you've done in the other cases below?

pchote commented on OpenRA.Game/GameRules/WeaponInfo.cs:
```diff
@@ -33,6 +33,15 @@ public class ProjectileArgs
 		public Target GuidedTarget;
 	}
 
+	public class WarheadArgs
```
I can see why we have this as its own distinct class instead of referencing `ProjectileArgs` directly (we have cases that don't start with a projectile), but having every use case manually copy over the fields adds a lot of duplication and potential headaches in the future if we want add more fields in the future.

We could solve this by adding an optional constructor
```csharp
public WarheadArgs(ProjectileArgs args)
{
	Weapon = args.Weapon;
	DamageModifiers = args.DamageModifiers;
	Source = args.Source;
	SourceActor = args.SourceActor;
	WeaponTarget = args.GuidedTarget;
}
```

and then cases like `AreaBeam` that do have a projectile but want to override with different values can do something like:
```csharp
var warheadArgs = new WarheadArgs(args)
{
	DamageModifiers = adjustedModifiers.ToArray(),
	Source = target
};
```

reaperrr replied:
Where would I have to put them? I'm getting a 'cannot put members directly in namespace' compile error if I just put them into WeaponInfo.

pchote replied:
```csharp
public class WarheadArgs
{
	public WeaponInfo Weapon;
	public int[] DamageModifiers;
	public WPos Source;
	public Actor SourceActor;
	public Target WeaponTarget;

	public WarheadArgs(ProjectileArgs args)
	{
		Weapon = args.Weapon;
		DamageModifiers = args.DamageModifiers;
		Source = args.Source;
		SourceActor = args.SourceActor;
		WeaponTarget = args.GuidedTarget;
	}

	// Default empty constructor for callers that want to initialize fields themselves
	public WarheadArgs() { }
}
```

pchote reviewed (commented):
Good idea, and a long time coming.

A couple of suggestions to reduce duplication and maintenance effort if we do need to add new fields in the future:

reaperrr commented on OpenRA.Mods.Common/Projectiles/AreaBeam.cs:
```diff
@@ -231,7 +232,17 @@ public void Tick(World world)
 				foreach (var a in actors)
 				{
 					var adjustedModifiers = args.DamageModifiers.Append(GetFalloff((args.Source - a.CenterPosition).Length));
```
Unrelated to this PR, but... am I reading this wrong, or does damage fall off the farther the hit actor is from the source actor, instead of the farther it is from the beam's center (which would be the correct behavior)?

pchote replied:
It looks like the desired behaviour according to #9357 was to have a certain section of the beam between `headPos` and `x` that deals maximum damage, then a linear falloff in damage between `x` and `y`, and then the remaining section from `y` to `tailPos` deal minimum damage.

reaperrr commented:
Updated.

pchote commented on OpenRA.Game/GameRules/WeaponInfo.cs:
```diff
@@ -177,17 +198,33 @@ public bool IsValidAgainst(FrozenActor victim, Actor firedBy)
 
 		/// <summary>Applies all the weapon's warheads to the target.</summary>
-		public void Impact(Target target, Actor firedBy, IEnumerable<int> damageModifiers)
+		public void Impact(Target impactTarget, WarheadArgs args)
```
Why rename `target` to `impactTarget`? IMO the `impact` is already implicit from the method name.

pchote commented on OpenRA.Game/GameRules/WeaponInfo.cs:
```diff
@@ -177,17 +198,33 @@ public bool IsValidAgainst(FrozenActor victim, Actor firedBy)
 		}
+
+		/// <summary>Applies all the weapon's warheads to the target. Only use for projectile-less, special-case impacts.</summary>
+		public void Impact(Target impactTarget, Actor firedBy)
```
Likewise here.

pchote commented on OpenRA.Mods.Cnc/Projectiles/IonCannon.cs:
```diff
@@ -45,9 +45,18 @@ public IonCannon(Player firedBy, WeaponInfo weapon, World world, WPos launchPos,
+				{
+					Weapon = weapon,
+					Source = target.CenterPosition,
+					SourceActor = firedBy.PlayerActor,
```
This should probably point at the advanced comm center / upgrade center that triggers the projectile.

reaperrr replied:
I faintly remember this being intentional though, probably to prevent actors from trying to retaliate to an 'attacker' in the middle of the enemy base.

pchote replied:
I hadn't considered that - makes sense.

pchote reviewed (commented):
A couple of minor thoughts, otherwise LGTM. I haven't tested this yet ingame.

pchote commented on OpenRA.Game/GameRules/WeaponInfo.cs:
```diff
@@ -33,6 +33,27 @@ public class ProjectileArgs
+	public class WarheadArgs
+	{
+		public WeaponInfo Weapon;
+		public int[] DamageModifiers;
```
This needs to be `public int[] DamageModifiers = { };` to avoid a crash when firing the ion cannon (and probably other things - the ion cannon was just the first thing I tested).

pchote commented on OpenRA.Mods.Common/Projectiles/NukeLaunch.cs:
```diff
@@ -140,7 +140,7 @@ void Explode(World world, bool removeProjectile)
 				return;
 
-			weapon.Impact(Target.FromPos(pos), firedBy.PlayerActor, Enumerable.Empty<int>());
+			weapon.Impact(Target.FromPos(pos), firedBy.PlayerActor);
```
This crashes because `firedBy.PlayerActor` doesn't have a `CenterPosition`. Manually constructing the `WarheadArgs` the same way as `IonCannonPower` should fix the issue.

pchote reviewed (commented):
Found a couple of crashes while testing:

reaperrr commented:
Updated.

teinarss commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/f7842f508f35100fade4a70172270e0ce3e5e51f...9e2aa3fd7b88e79785589245fa059a3a32e2faf6
